Add support for X11 primary selection

Desour opened the pull request:
## Description

X11 has a so-called primary selection, which you can use by marking text and middle-clicking elsewhere to copy the marked text.

There are 3 new API functions in `SDL_clipboard.h`, which work exactly like their clipboard equivalents.
The code is also essentially just a duplication of the clipboard code. The diffs make this look larger than it is. When reviewing, compare the functions with their clipboard equivalent (directly above). (I'm open for suggestions to reduce code duplication.)

~~`SDL_SetClipboardText` still writes into clipboard *and* primary selection. I thought it's better to keep it like this for backwards compatibility. And it's not disturbing if you use clipboard and primary selection normally (you can only copy marked text into the clipboard).~~

Style fixes by clang-format were only applied very conservatively, because it wants to break alignment in `SDL_events.h`, and most suggestions just lead to inconsistency with neighbouring old code.

## Test Instructions

* Run the tests (just a copy of the clipboard tests): `$ ./test/testautomation --filter Clipboard`
* Build and run this small application:
<details>

```C
#include <SDL.h>
#include <unistd.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void print_error(const char *where)
{
	const char *errstr = SDL_GetError();
	if (errstr == NULL || errstr[0] == '\0')
		return;
	fprintf(stderr, "SDL Error after '%s': %s\n", where, errstr);
	SDL_ClearError();
}

int main()
{
	char text_buf[256];

	srand(time(NULL));

	SDL_Init(SDL_INIT_VIDEO);
	print_error("SDL_INIT()");
	SDL_Window *window = SDL_CreateWindow("Primary Selection Test", SDL_WINDOWPOS_UNDEFINED,
			SDL_WINDOWPOS_UNDEFINED, 400, 400, SDL_WINDOW_SHOWN);
	print_error("SDL_CreateWindow()");
	SDL_Renderer *renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED);
	print_error("SDL_CreateRenderer()");

	bool quit = false;
	unsigned int do_render = 0;
	while (!quit) {
		SDL_Event event;
		while (SDL_PollEvent(&event)) {
			print_error("SDL_PollEvent()");
			switch (event.type) {
			case SDL_QUIT: {
				quit = true;
				break;
			} case SDL_KEYDOWN: {
				switch (event.key.keysym.sym) {
				case SDLK_ESCAPE:
				case SDLK_q:
					quit = true;
					break;
				case SDLK_c:
					snprintf(text_buf, sizeof(text_buf), "foo%d", rand());
					SDL_SetClipboardText(text_buf);
					print_error("SDL_SetClipboardText()");
					printf("clipboard: set_to=\"%s\"\n", text_buf);
					break;
				case SDLK_v: {
					printf("clipboard: has=%d, ", SDL_HasClipboardText());
					print_error("SDL_HasClipboardText()");
					char *text = SDL_GetClipboardText();
					print_error("SDL_GetClipboardText()");
					printf("text=\"%s\"\n", text);
					SDL_free(text);
					break;
				} case SDLK_d:
					snprintf(text_buf, sizeof(text_buf), "bar%d", rand());
					SDL_SetPrimarySelectionText(text_buf);
					print_error("SDL_SetPrimarySelectionText()");
					printf("primselec: set_to=\"%s\"\n", text_buf);
					break;
				case SDLK_f: {
					printf("primselec: has=%d, ", SDL_HasPrimarySelectionText());
					print_error("SDL_HasPrimarySelectionText()");
					char *text = SDL_GetPrimarySelectionText();
					print_error("SDL_GetPrimarySelectionText()");
					printf("text=\"%s\"\n", text);
					SDL_free(text);
					break;
				} default:
					break;
				}
				break;
			} default: {
				break;
			}}
		}
		// create less noise with WAYLAND_DEBUG=1
		if (do_render == 0) {
			SDL_RenderPresent(renderer);
			print_error("SDL_RenderPresent()");
		}
		do_render += 1;
		usleep(12000);
	}

	SDL_DestroyRenderer(renderer);
	SDL_DestroyWindow(window);
	SDL_Quit();
	print_error("quit");
	return 0;
}
```

</details>

* Use c,v,d,f to get and set the clipboard and primary selection.
* Mark text and middle-click also in other applications.
* For wayland under x:
  * `$ mutter --wayland --no-x11 --nested`
  * `$ XDG_SESSION_TYPE=wayland SDL_VIDEODRIVER=wayland ./<path_to_test_appl_binary>`

## Existing Issue(s)

Found none.

Desour commented on src/video/x11/SDL_x11clipboard.c:
```diff
@@ -200,13 +193,13 @@ X11_GetClipboardText(_THIS)
                 videodata->selection_waiting = SDL_FALSE;
-                SDL_SetError("Clipboard timeout");
-                /* We need to set the clipboard text so that next time we won't
+                SDL_SetError("Selection timeout");
```
Is changing error message text considered a breaking change?

flibitijibibo replied:
I don't think so, no - they're generally meant for logging and not any specific functionality.

flibitijibibo commented:
Someone on the X side will need to review this, but a good head-start on adding this would be to add at least one more backend to this (it's not a hard policy AFAIK, but personally I like Vulkan's rule of requiring multiple implementations before finalizing specs). I believe there's a Wayland protocol for primary selection and I would definitely be able to review that; not sure what the other OSes offer.

slouken commented:
Windows and macOS don't have the concept of a separate selection from the clipboard. I'm not sure that this has much cross-platform value.

Desour commented:
I'll try to add an implementation for wayland (according to [this docs](https://wayland.app/protocols/primary-selection-unstable-v1)). It might take a while though, as I'm new to wayland land.

icculus commented:
I'd be inclined to say we don't want to add this.

flibitijibibo commented:
Looks like this might matter for minetest: https://github.com/minetest/irrlicht/pull/132

That's up to their team though - also for all I know this can be implemented without SDL internals, not sure. Wayland's registry system makes it not-too-scary, not sure about X.

slouken commented:
I think this makes sense as a dynamic hint. There’s a single clipboard API entry point, but applications can decide which selection they care about, and it doesn’t affect platforms that don’t support that.

slouken commented:
This also helps with the code duplication, you just need a single function to pick the selection based on the hint (or lack thereof)

I would be comfortable with adding this change as a hint without a Wayland implementation (that’s what hints are for) and that can come later.

Desour commented:
> Looks like this might matter for minetest

Not more than for other applications that have text input fields. :-)


> I think this makes sense as a dynamic hint. There’s a single clipboard API entry point, but applications can decide which selection they care about, and it doesn’t affect platforms that don’t support that.

That would mean that on not affected platforms, the clipboard text is overridden whenever the user marks some text, no?

>This also helps with the code duplication, you just need a single function to pick the selection based on the hint (or lack thereof)

The code duplication also happens *inside* the picked functions. (This also seems to be the case for the wayland primary selection btw., the `zwp_primary_selection_.*` stuff seems to work extremely similar to the `wl_data_.*` stuff, just with different type and function names everywhere.)

Desour commented:
I've implemented it for wayland now. (Mostly copy-paste of the `wl_data` stuff, replacing the names.)
It should work in theory, but I couldn't test it yet, because weston, which can run under xorg via its x11 backend, doesn't support primary selection (see also <https://wayland.app/protocols/primary-selection-unstable-v1#compositor-support>).
(Note: I've updated the test application, because wayland doesn't show a window if you don't render anything.)

flibitijibibo commented:
Wayland portion looks okay to me. I would definitely be interested in the hint idea if possible, since as Sam mentioned nobody else exposes this idea (I couldn't find any other examples myself).

Even if it's not ideal for full multiplat builds, this _definitely_ seems like a good idea to help X applications (like minetest) migrate to Wayland while also retaining X support, and get all the other SDL benefits along the way.

Desour commented:
Maybe I'm getting something wrong, but the API would look like with hints, right:
```c
SDL_SetHint(SDL_HINT_CLIPBOARD_SELECTION, "primary");
SDL_SetClipboardText("some marked text");
```
And on MacOS for example, this would write `"some marked text"` into the clipboard.
So, to avoid this, the natural thing to do for an application would be to define a function like this:
```c
void set_primary_selection_text(const char *text)
{
#if X11
SDL_SetHint(SDL_HINT_CLIPBOARD_SELECTION, "primary");
SDL_SetClipboardText(text);
SDL_SetHint(SDL_HINT_CLIPBOARD_SELECTION, "clipboard");
#endif
}
```
Now I'm wondering, how this would be beneficial.

Desour commented:
Tested using mutter. Primary selection now works as good as the clipboard.

However, for both (and also on master), mutter is complaining about broken pipes, and copying to a second sdl application window fails about every 2nd time:
```
(mutter:27161): libmutter-WARNING **: 02:30:14.100: Could not fetch selection data: Fehler beim Schreiben in den Dateideskriptor: Datenübergabe unterbrochen (broken pipe)
```
(The german part says "Error at writing to file descriptor: file transmission aborted".)
Is this a known issue (haven't found any issue)?

flibitijibibo commented:
Wish I knew off the top of my head... the only other people who've worked on clipboard/data marshaling recently was Sam and @cgutman. `WAYLAND_DEBUG=1` might reveal something in between all the usual spam?

meyraud705 commented:
> I think this makes sense as a dynamic hint. There’s a single clipboard API entry point, but applications can decide which selection they care about, and it doesn’t affect platforms that don’t support that.

Primary selection is a second clipboard independent of regular clipboard with its own shortcut (copy:selection, paste: middle mouse button). It make sense to have new API for this and a second clipboard storage. 

If an application does not call the new API, the feature won't be supported, so you don't need the hint.

For platform that do not support this feature, I like the current implementation that make primary selection internal to the application.

slouken commented:
Okay, I'm inclined to accept this, once it's rebased to fix the dynapi conflicts and the wayland error is solved.

@icculus, thoughts?

Desour commented:
Rebased.

Desour commented:
Btw., something that might maybe make sense to control via hints is:
* Whether `setClipboardText` also sets primary selection on X11.
* The behaviour of platforms that do not support primary selection.
* The behaviour on wayland if the compositor does not support primary selection.

But I don't know if that would make the API just overly complex.

Desour commented:
I've fixed the broken pipe issues. See commit descriptions for details. (Here's how I measured the latency, FYI: <https://github.com/Desour/SDL/commit/736803ccc2ec5dfc656494cbf2c853d0e5c14847>)

flibitijibibo reviewed (approved):
With the Wayland fix in, gets a +1 from me.

icculus commented:
I still think this isn't something we want.

We're adding an API for a feature that means literally nothing except on two platforms (one of them meant to obsolete the other), which already have functioning clipboards through the SDL API.

The broken pipe timing commits are pretty scary, too.

meyraud705 commented:
Personally, I use primary selection a lot. For "modern" applications and websites that require a mouse to navigate, it is convenient to have a clipboard accessible from the mouse. However, I almost never used any type of copy/paste in game. Maybe people that plays games that would benefit from this feature could explain better than me why it need to be included in SDL.

I'd assume discoverability of primary selection is low. I wonder how many Linux user know about it and how many use it. Plus, game-developer that uses windows will probably not implement this feature even if it becomes available in SDL. Maybe adding an explanation of what primary selection is in the documentation would help.

About the pipe timeout: comment above say "infinite loop of PumpEvents", does commit 8bf32e1 fix this?

If it gets merged, I think it would be better if both clipboard where fully independent (`setClipboardText` does not sets primary selection). This make a difference if there is button that allow to copy a field without selecting it. An example of such button can be found on Github: in the commit list there is a "copy the full SHA" button.

flibitijibibo commented:
Basically that timeout define is a bandaid on something I couldn't fully investigate at the time - all I could gather at the time is that if your wait time for data events was longer than presentation, you would get yourself in a situation where both presentation and events would take _way_ too long and eventually a pipe would break. This patch, if nothing else, showed that the wait time works both ways when it comes to avoiding broken pipes.

Also: While games aren't _too_ text heavy, the time when you _need_ text input is basically an all-or-nothing situation, and at least on FNA's end we've had developers hit the wall more often than any other subsystem... so when I see something where SDL would be a blocker for full text support, I now lean towards trying to include it because it's an area that, relative to other parts of SDL, our feature support isn't _nearly_ as thorough and that could bite us later on since there are only so many ways to support these things without just removing SDL from the equation.

slouken commented:
> Also: While games aren't _too_ text heavy, the time when you _need_ text input is basically an all-or-nothing situation, and at least on FNA's end we've had developers hit the wall with text more often than any other subsystem... so when I see something where SDL would be a blocker for full text support, I now lean towards trying to include it because it's an area that, relative to other parts of SDL, our feature support isn't _nearly_ as thorough and that could bite us later on since there are only so many ways to support these things without just removing SDL from the equation.

FNA games are used to using the Windows clipboard, so the concept of primary selection is foreign to them, right?

flibitijibibo commented:
> > Also: While games aren't _too_ text heavy, the time when you _need_ text input is basically an all-or-nothing situation, and at least on FNA's end we've had developers hit the wall with text more often than any other subsystem... so when I see something where SDL would be a blocker for full text support, I now lean towards trying to include it because it's an area that, relative to other parts of SDL, our feature support isn't _nearly_ as thorough and that could bite us later on since there are only so many ways to support these things without just removing SDL from the equation.
> 
> FNA games are used to using the Windows clipboard, so the concept of primary selection is foreign to them, right?

To a certain degree yes, but it's worth noting that XNA didn't have _any_ specification for text, since Xbox 360 didn't really have such a concept (even though it had the OSK and everything), so we had to carve that out ourselves. Problem is, most of what we carve out is just exposing SDL, and so we regularly find ourselves tacking on more and more to our extension to fill in for various gaps. Some gaps are admittedly us not exposing an SDL function (the big one was input rectangles, that was my bad for sure), some of it is us not exposing weird details like control keys (i.e. Home/End), but to a small degree some of it is us simply not having any real knowledge of how to expose very specific types of text input, and this in my mind is one such case. I would not be surprised if there were more examples, I imagine many of them will be coming from people trying to submit text on Steam Deck.

FNA's text input extension spec can be found [here](https://github.com/FNA-XNA/FNA/wiki/5:-FNA-Extensions#textinputext), but just by looking at it now I realize we haven't documented everything we expose... so here's the [frontend](https://github.com/FNA-XNA/FNA/blob/master/src/Input/TextInputEXT.cs) and [backend](https://github.com/FNA-XNA/FNA/blob/master/src/FNAPlatform/SDL2_FNAPlatform.cs).

Desour commented:
There's nothing left to do in this PR, btw..
Is there something I can do to speed up the adoption process?

slouken commented:
@icculus and I are going to review for the next milestone.

slouken commented:
Given that this has a working internal implementation for non-Unix platforms and two implementations for Unix platforms, I'm inclined to accept this PR. The API is entirely optional, so shouldn't affect applications that don't care about it.

Desour commented:
Thank you! ^^
